### DISCUSSION TITLE
Unable to create simulated chargers

### DISCUSSION THREAD

**Discussion by kaanTgun** [SAP/e-mobility-charging-stations-simulator#1238] [Q&A] [answered]
Hello have spent a few hours trying to get the simulation working but I was not successful. I'm working on the master branch  
I feel like I'm missing something. I would greatly appreciate it if you could help me with this. 

Thanks

I did follow the readme, (installing the node packages etc.) It was straight forward. I was also able to get the UI working. However my simulator is not able to simulate the charge(s).

my zsh logs 
```
>  e-mobility-charging-stations-simulator git:(main) ✗ pnpm start   
> e-mobility-charging-stations-simulator@2.0.1 start e-mobility-charging-stations-simulator
> pnpm build && cross-env NODE_ENV=production node dist/start.js
> e-mobility-charging-stations-simulator@2.0.1 build e-mobility-charging-stations-simulator
> pnpm esbuild
> e-mobility-charging-stations-simulator@2.0.1 esbuild e-mobility-charging-stations-simulator
> pnpm build-requirements && node scripts/bundle.js
> e-mobility-charging-stations-simulator@2.0.1 build-requirements e-mobility-charging-stations-simulator
> node scripts/build-requirements.js
> 
> Building in production mode
> Build time: 285.169ms
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:ChangeAvailabilityRequest#/properties/type" (strictTypes)
> Charging stations simulator 2.0.1 started with 1 configured and 0 provisioned charging station(s) from 1 charging station template(s) and 1 worker(s) concurrently running in 'workerSet' mode (1 charging station(s) per worker)
> Worker set/pool information: {
>   elementsExecuting: 1,
>   elementsPerWorker: 1,
>   size: 1,
>   started: true,
>   type: 'set',
>   version: '1.0.1',
>   worker: 'thread'
> }
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:ClearChargingProfileRequest#/properties/chargingProfilePurpose" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:GetCompositeScheduleRequest#/properties/chargingRateUnit" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:RemoteStartTransactionRequest#/properties/chargingProfile/properties/chargingProfilePurpose" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:RemoteStartTransactionRequest#/properties/chargingProfile/properties/chargingProfileKind" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:RemoteStartTransactionRequest#/properties/chargingProfile/properties/recurrencyKind" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:RemoteStartTransactionRequest#/properties/chargingProfile/properties/chargingSchedule/properties/chargingRateUnit" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:ResetRequest#/properties/type" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:SetChargingProfileRequest#/properties/csChargingProfiles/properties/chargingProfilePurpose" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:SetChargingProfileRequest#/properties/csChargingProfiles/properties/chargingProfileKind" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:SetChargingProfileRequest#/properties/csChargingProfiles/properties/recurrencyKind" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:SetChargingProfileRequest#/properties/csChargingProfiles/properties/chargingSchedule/properties/chargingRateUnit" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:TriggerMessageRequest#/properties/requestedMessage" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:AuthorizeResponse#/properties/idTagInfo/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:BootNotificationResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:DataTransferResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:StartTransactionResponse#/properties/idTagInfo/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:StopTransactionResponse#/properties/idTagInfo/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:CancelReservationResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:ChangeAvailabilityResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:ChangeConfigurationResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:ClearCacheResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:ClearChargingProfileResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:DataTransferResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:GetCompositeScheduleResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:GetCompositeScheduleResponse#/properties/chargingSchedule/properties/chargingRateUnit" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:RemoteStartTransactionResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:RemoteStopTransactionResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:ReserveNowResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:ResetResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:SetChargingProfileResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:TriggerMessageResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:UnlockConnectorResponse#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:DiagnosticsStatusNotificationRequest#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:FirmwareStatusNotificationRequest#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:MeterValuesRequest#/properties/meterValue/items/properties/sampledValue/items/properties/context" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:MeterValuesRequest#/properties/meterValue/items/properties/sampledValue/items/properties/format" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:MeterValuesRequest#/properties/meterValue/items/properties/sampledValue/items/properties/measurand" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:MeterValuesRequest#/properties/meterValue/items/properties/sampledValue/items/properties/phase" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:MeterValuesRequest#/properties/meterValue/items/properties/sampledValue/items/properties/location" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:MeterValuesRequest#/properties/meterValue/items/properties/sampledValue/items/properties/unit" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:StatusNotificationRequest#/properties/errorCode" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:StatusNotificationRequest#/properties/status" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:StopTransactionRequest#/properties/reason" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:StopTransactionRequest#/properties/transactionData/items/properties/sampledValue/items/properties/context" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:StopTransactionRequest#/properties/transactionData/items/properties/sampledValue/items/properties/format" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:StopTransactionRequest#/properties/transactionData/items/properties/sampledValue/items/properties/measurand" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:StopTransactionRequest#/properties/transactionData/items/properties/sampledValue/items/properties/phase" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:StopTransactionRequest#/properties/transactionData/items/properties/sampledValue/items/properties/location" (strictTypes)
> strict mode: missing type "object" for keyword "additionalProperties" at "urn:OCPP:1.6:2019:12:StopTransactionRequest#/properties/transactionData/items/properties/sampledValue/items/properties/unit" (strictTypes)

```

combined-2024-12-01.log 

> info: 12/1/2024, 2:28:41 PM CS-SIEMENS | Open OCPP connection to URL ws://localhost:8010/OCPP16/5be7fb271014d90008992f06/CS-SIEMENS
info: 12/1/2024, 2:28:41 PM Bootstrap | Bootstrap.workerEventAdded: Charging station CS-SIEMENS (hashId: 256627d21bf834004d09a1c7b438a328ba667a2742882df6c2ae20764c8208237e1f0e5b0cf44f46fc836bb9046932f6) added (0 added from 1 configured and 0 provisioned charging station(s))
info: 12/1/2024, 2:28:41 PM Bootstrap | Bootstrap.workerEventStarted: Charging station CS-SIEMENS (hashId: 256627d21bf834004d09a1c7b438a328ba667a2742882df6c2ae20764c8208237e1f0e5b0cf44f46fc836bb9046932f6) started (1 started from 0 added charging station(s))
error: 12/1/2024, 2:28:41 PM CS-SIEMENS | WebSocket error: {"code":"ECONNREFUSED","stack":"AggregateError [ECONNREFUSED]: \n    at internalConnectMultiple (node:net:1121:18)\n    at afterConnectMultiple (node:net:1688:7)"}
error: 12/1/2024, 2:28:41 PM CS-SIEMENS | WebSocket abnormally closed with status 'Abnormal Closure' and reason ''
error: 12/1/2024, 2:28:41 PM CS-SIEMENS | WebSocket connection retry in 30000ms, timeout 29000ms
error: 12/1/2024, 2:29:11 PM CS-SIEMENS | WebSocket connection retry #1
info: 12/1/2024, 2:29:11 PM CS-SIEMENS | Open OCPP connection to URL ws://localhost:8010/OCPP16/5be7fb271014d90008992f06/CS-SIEMENS
error: 12/1/2024, 2:29:11 PM CS-SIEMENS | WebSocket error: {"code":"ECONNREFUSED","stack":"AggregateError [ECONNREFUSED]: \n    at internalConnectMultiple (node:net:1121:18)\n    at afterConnectMultiple (node:net:1688:7)"}
error: 12/1/2024, 2:29:11 PM CS-SIEMENS | WebSocket abnormally closed with status 'Abnormal Closure' and reason ''
error: 12/1/2024, 2:29:11 PM CS-SIEMENS | WebSocket connection retry in 30000ms, timeout 29000ms
error: 12/1/2024, 2:29:41 PM CS-SIEMENS | WebSocket connection retry #2
info: 12/1/2024, 2:29:41 PM CS-SIEMENS | Open OCPP connection to URL ws://localhost:8010/OCPP16/5be7fb271014d90008992f06/CS-SIEMENS
error: 12/1/2024, 2:29:41 PM CS-SIEMENS | WebSocket error: {"code":"ECONNREFUSED","stack":"AggregateError [ECONNREFUSED]: \n    at internalConnectMultiple (node:net:1121:18)\n    at afterConnectMultiple (node:net:1688:7)"}
error: 12/1/2024, 2:29:41 PM CS-SIEMENS | WebSocket abnormally closed with status 'Abnormal Closure' and reason ''
error: 12/1/2024, 2:29:41 PM CS-SIEMENS | WebSocket connection retry in 30000ms, timeout 29000ms
error: 12/1/2024, 2:30:11 PM CS-SIEMENS | WebSocket connection retry #3
info: 12/1/2024, 2:30:11 PM CS-SIEMENS | Open OCPP connection to URL ws://localhost:8010/OCPP16/5be7fb271014d90008992f06/CS-SIEMENS
error: 12/1/2024, 2:30:11 PM CS-SIEMENS | WebSocket error: {"code":"ECONNREFUSED","stack":"AggregateError [ECONNREFUSED]: \n    at internalConnectMultiple (node:net:1121:18)\n    at afterConnectMultiple (node:net:1688:7)"}
error: 12/1/2024, 2:30:11 PM CS-SIEMENS | WebSocket abnormally closed with status 'Abnormal Closure' and reason ''
error: 12/1/2024, 2:30:11 PM CS-SIEMENS | WebSocket connection retry in 30000ms, timeout 29000ms

error-2024-12-01.log

> error: 12/1/2024, 2:28:41 PM CS-SIEMENS | WebSocket error: {"code":"ECONNREFUSED","stack":"AggregateError [ECONNREFUSED]: \n    at internalConnectMultiple (node:net:1121:18)\n    at afterConnectMultiple (node:net:1688:7)"}
error: 12/1/2024, 2:28:41 PM CS-SIEMENS | WebSocket abnormally closed with status 'Abnormal Closure' and reason ''
error: 12/1/2024, 2:28:41 PM CS-SIEMENS | WebSocket connection retry in 30000ms, timeout 29000ms
error: 12/1/2024, 2:29:11 PM CS-SIEMENS | WebSocket connection retry #1
error: 12/1/2024, 2:29:11 PM CS-SIEMENS | WebSocket error: {"code":"ECONNREFUSED","stack":"AggregateError [ECONNREFUSED]: \n    at internalConnectMultiple (node:net:1121:18)\n    at afterConnectMultiple (node:net:1688:7)"}
error: 12/1/2024, 2:29:11 PM CS-SIEMENS | WebSocket abnormally closed with status 'Abnormal Closure' and reason ''
error: 12/1/2024, 2:29:11 PM CS-SIEMENS | WebSocket connection retry in 30000ms, timeout 29000ms
error: 12/1/2024, 2:29:41 PM CS-SIEMENS | WebSocket connection retry #2
error: 12/1/2024, 2:29:41 PM CS-SIEMENS | WebSocket error: {"code":"ECONNREFUSED","stack":"AggregateError [ECONNREFUSED]: \n    at internalConnectMultiple (node:net:1121:18)\n    at afterConnectMultiple (node:net:1688:7)"}
error: 12/1/2024, 2:29:41 PM CS-SIEMENS | WebSocket abnormally closed with status 'Abnormal Closure' and reason ''
error: 12/1/2024, 2:29:41 PM CS-SIEMENS | WebSocket connection retry in 30000ms, timeout 29000ms
error: 12/1/2024, 2:30:11 PM CS-SIEMENS | WebSocket connection retry #3
error: 12/1/2024, 2:30:11 PM CS-SIEMENS | WebSocket error: {"code":"ECONNREFUSED","stack":"AggregateError [ECONNREFUSED]: \n    at internalConnectMultiple (node:net:1121:18)\n    at afterConnectMultiple (node:net:1688:7)"}
error: 12/1/2024, 2:30:11 PM CS-SIEMENS | WebSocket abnormally closed with status 'Abnormal Closure' and reason ''
error: 12/1/2024, 2:30:11 PM CS-SIEMENS | WebSocket connection retry in 30000ms, timeout 29000ms
error: 12/1/2024, 2:30:41 PM CS-SIEMENS | WebSocket connection retry #4
error: 12/1/2024, 2:30:41 PM CS-SIEMENS | WebSocket error: {"code":"ECONNREFUSED","stack":"AggregateError [ECONNREFUSED]: \n    at internalConnectMultiple (node:net:1121:18)\n    at afterConnectMultiple (node:net:1688:7)"}
error: 12/1/2024, 2:30:41 PM CS-SIEMENS | WebSocket abnormally closed with status 'Abnormal Closure' and reason ''
error: 12/1/2024, 2:30:41 PM CS-SIEMENS | WebSocket connection retry in 30000ms, timeout 29000ms
error: 12/1/2024, 2:31:11 PM CS-SIEMENS | WebSocket connection retry #5

<img width="1469" alt="Screenshot 2024-12-01 at 2 54 24 PM" src="https://github.com/user-attachments/assets/4e823237-f52f-4f48-9f8f-748880cd26b3">

This is what my src/assets/config.json file looks like.

>
```
 {

  "supervisionUrls": ["ws://localhost:8010/OCPP16/5be7fb271014d90008992f06"],
  "supervisionUrlDistribution": "round-robin",
  "log": {
    "file": "logs/combined.log",
    "errorFile": "logs/error.log",
    "maxFiles": 7,
    "statisticsInterval": 0
  },
  "worker": {
    "processType": "workerSet"
  },
  "performanceStorage": {
    "enabled": true,
    "type": "none"
  },
  "uiServer": {
    "enabled": true,
    "type": "ws",
    "authentication": {
      "enabled": false,
      "type": "protocol-basic-auth",
      "username": "admin",
      "password": "admin"
    }
  },
  "stationTemplateUrls": [
    {
      "file": "siemens.station-template.json",
      "numberOfStations": 1
    }
  ]
}
```

ui/web/public/config.json
```
{
  "uiServer": {
    "host": "localhost",
    "port": 8080,
    "protocol": "ui",
    "version": "0.0.1",
    "authentication": {
      "enabled": true,
      "type": "protocol-basic-auth",
      "username": "admin",
      "password": "admin"
    }
  }
}
```

I havent made any other changes and I'm working on a M2 Mac.

**Comment by jerome-benoit** [comment]
There's no OCPP server started and listening on localhost port 8010. Running a client of a client/server design requires some minimum knowledge about client/server architecture.

**Reply by kaanTgun** [reply]
Hi Jerome thank you for your response. 
The part that is not clear to me is the communication between the UIServer and the simulated charge points. I was under the assumption that UIServer was setup to act as a simple OCPP server. But I believe its just a tool to communicate actions with the OCPP server and the simulated charge points. So to run this end-to-end I need to run a OCPP server like Steve and use its endpoint by setting `supervisionUrls` at src/assets/config.json. 

Am I correct ?

Thank you

**Reply by jerome-benoit** [reply, marked as answer]
The documentation is implicit about OCPP server requirement and assumes the user has at minimum read and understood the OCPP protocol specs.
